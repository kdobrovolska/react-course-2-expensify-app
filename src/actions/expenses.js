import uuid from 'uuid';
import database from '../firebase/firebase'

/* export const AddExpense=({description='',note='',amount=1,createdAt=0}={})=>(
    {
        type:'ADD_EXPENCE',
        expense:{id:uuid(),description,note,amount,createdAt}
    }
); */
export const AddExpense=(expense)=>(
{
    type:'ADD_EXPENCE',
    expense
});

export const startAddExpense=(expenseData={})=>{
    return  (dispatch) => {
        const {
            description=0,
            note='',
            amount=0,
            createAt=0
        } = expenseData;
        const expense={description, note, amount, createAt};
        return database.ref('expenses').push(expense).then((ref) => { 
            const pp='expenses'+'/'+ref.key;
            console.log('push.then::'+database.ref(pp));
            dispatch(AddExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}
export const RemoveExpense=(id)=>(
    {
        type:'REMOVE_EXPENCE',
        expenseID:id
    }
);
export const EditExpense=(id, updates={})=>(
    {
        type:'EDIT_EXPENCE',
        expenseID:id,
        expense: updates
    }
);