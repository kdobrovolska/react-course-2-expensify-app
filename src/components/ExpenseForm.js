import React from 'react';
import moment from 'moment';
const now =moment();
console.log(now);
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'; //go to app.js

export default class ExpenseForm extends React.Component
{
    
    constructor(props)
    {
        
      //  alert('conzstructor');
        super(props);
        this.state={description:'',
            note:'', 
            amount:'',
            createdAt:moment(),
            calendarFocused:false,
            error:''
            };
        
        if(!!props.expense)
        {
          this.state =  {...this.state,...props.expense,createdAt:moment(props.expense.CreateAt)};
           
        }
    }
    
    DescriptionChanged=(e)=>
    {
        
        const description=e.target.value;
      //  alert(description);
        this.setState(()=>({description}));
    }
    NoteChanged=(e)=>
    {
        e.persist();
         this.setState(()=>({note:e.target.value}));
    }
    AmountChanged=(e)=>
    {
        
        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    onDateChange=(createdAt) =>{
        if(createdAt)
            this.setState(()=>({createdAt}));
    }
    onFocusChange1=({focused})=>
    {
       this.setState(()=>({calendarFocused:focused}));
    }
    onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount)
        {
            this.setState(()=>({error:'Please provide description and amount'}));
       
        }else{
            this.setState(()=>({error:''}));
            this.props.submit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            });
        }
    };
    render()
    {
        let obj=this.state;
     /*   if(!!this.props.expense)
        {
           obj=  {...obj,...this.props.expense,createdAt:moment(this.props.expense.CreateAt)};
        }*/

     //   alert('render');
      return <form className='form' onSubmit={ this.onSubmit}>
       {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input type='text' 
            className='text-input' 
            placeholder='Description' 
            autoFocus
            value={obj.description}
            onChange={this.DescriptionChanged}
            />
        <label id='errDescription' className='err'></label>
        <input type='text' 
            className='text-input' 
            placeholder='Amount' 
            value={obj.amount}
            onChange={this.AmountChanged}
            />
            
         <SingleDatePicker 
            date={obj.createdAt}
            onDateChange={this.onDateChange}
            focused ={this.state.calendarFocused}
            onFocusChange ={this.onFocusChange1}
            numberOfMonths={1}
            isOutsideRange={()=>{ return false;}}
            />
        <p></p>
        <textarea placeholder='Add note' 
            className='textarea'
            value={obj.note}
            onChange={this.NoteChanged}
        ></textarea>
        <div>
            <button className='button1'>Save Expense</button>
        </div>
  
        </form> ;
    }
}