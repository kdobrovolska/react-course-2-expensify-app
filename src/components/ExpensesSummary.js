import React from 'react'
import {connect} from 'react-redux';
import numeral  from 'numeral'
import visibleExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expensesTotal'


export class ExpensesSummary extends React.Component
{
    render()
    {
       return <div>
        <h2>Expenses Count : {this.props.expensesCount} </h2>
        <h2>Expenses Total : {numeral(this.props.expensesTotal/100).format('$0,0.00')} </h2>
        </div>
    }
}
const MapStoreToProps=(state)=>
{
    return { 
        expensesCount:visibleExpenses(state.expenses,state.filters).length,
        expensesTotal:expensesTotal(visibleExpenses(state.expenses,state.filters))
    };
}
export default connect(MapStoreToProps)(ExpensesSummary);