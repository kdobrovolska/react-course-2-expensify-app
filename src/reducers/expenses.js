const expenseReducerDefaultState=[];

export default (state=expenseReducerDefaultState,action)=>{
    let ar;
    switch(action.type){
        case 'ADD_EXPENCE':
            return [...state,action.expense];
        case 'REMOVE_EXPENCE':
              return state.filter(({id})=>{return id!=action.expenseID});
        case 'EDIT_EXPENCE':
     //   alert('id='+action.expenseID);
           return state.map((item)=>{
               if(item.id===action.expenseID)
               {
                  // alert(action.expense.description);
                   return {...item,...action.expense, id:action.expenseID}
               }
               else
               {
                   return item;
               }
            })
        case 'SORT_BY_AMOUNT':
          ar=[...state];
          ar.sort((item1,item2)=>( item1.amount-item2.amount));
          return ar;
        case 'SORT_BY_DATE':
          ar=[...state];
          ar.sort((item1,item2)=>( item1.createdAt-item2.createdAt));
          return ar;
        default:
         return state;
    }
};