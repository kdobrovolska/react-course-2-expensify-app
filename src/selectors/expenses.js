import moment from 'moment'
export default (expenses,{startDate,endDate,text, sortBy})=>
{
  /*  const ar= [];
    expenses.map((item)=>{
        
        const isStartDate= (typeof startDate != 'number' ) || item.createdAt>=startDate;
        const isEndDate= (typeof endDate != 'number' ) || item.createdAt<=endDate;
        const isText = true; //(typeof endDate != 'string') || item.description.toLowerCase().indexOf(text.toLowerCase())>=0;
     //   alert(isStartDate+' , '+isEndDate+' , '+ isText );
      //  alert(startDate+' , '+endDate+' , '+ text +'=='+item.createdAt);
        if( isStartDate && isEndDate && isText)
        {
            ar.push(item);
        }
    });
   
    console.log('count='+expenses.length, ar.length);
   return ar;*/
   return expenses.filter((item)=>{
        const createdAtMoment=moment(item.createdAt);
        const isStartDate=  startDate? startDate.isSameOrBefore(createdAtMoment, 'day'):true;
        const isEndDate= endDate? endDate.isSameOrAfter(createdAtMoment, 'day'):true;
        const isText = (typeof text != 'string') || item.description.toLowerCase().includes(text.toLowerCase());
        return isStartDate && isEndDate && isText;
    }).sort((a,b)=>{
        if(sortBy.toLowerCase()==='date')
            return  (a.createdAt-b.createdAt);
        if(sortBy.toLowerCase()==='amount')
            return  (a.amount-b.amount);
  
    }); 
    
}