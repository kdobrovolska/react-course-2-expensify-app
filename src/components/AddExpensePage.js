import React from 'react';

import {connect} from 'react-redux';
import {AddExpense} from '../actions/expenses.js';
import ExpenseForm from './ExpenseForm';

/* const AddExpensePage = (props) => (
  <div>
    <h1> Add Form:</h1>
        <ExpenseForm 
           submit={(obj)=>  {
            console.log(obj);
            props.dispatch(AddExpense(obj));
            props.history.push('/');
        } } 
      />
  </div>
);
export default connect()(AddExpensePage); */


export class  AddExpensePage extends React.Component {
  onSubmit=(expense)=>{
      this.props.AddExpense(expense);
      this.props.history.push('/');
  }

  render()
  {  return   <div>
    <h1> Add Form:</h1>
        <ExpenseForm  submit={ this.onSubmit} />
    </div>;
  }
}
const MapDispatchToProps =(dispatch)=>{
 return { AddExpense:(expense)=>dispatch(AddExpense(expense))}
}
export default connect(undefined,MapDispatchToProps)(AddExpensePage);