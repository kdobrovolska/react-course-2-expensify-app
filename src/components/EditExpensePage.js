import React from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import ExpenseForm from './ExpenseForm';
import {startEditExpense, RemoveExpense,startRemoveExpense} from '../actions/expenses.js';

export class EditExpensePage extends React.Component{

   submit=(obj)=>  {
          this.props.startEditExpense(this.props.expense.id,obj);
          this.props.history.push('/');
        }

   onClick=()=>{
          this.props.startRemoveExpense(this.props.expense.id);
          this.props.history.push('/');
        }

    render()
    {
        return <div>
          <div className='page-header'>
            <div className='content-container'>
                <h1 className='page-header__title'>Edit Expense </h1>
            </div>
           </div>
           <div className='content-container'>
              <ExpenseForm expense={this.props.expense}
                submit={this.submit } />
              <button className='button1 button1__secondary' onClick={ this.onClick}>Remove Expense</button>
          </div>
        </div>;
    }
};

const MapStoreToProps=(state,props)=>
{
     return { expense:state.expenses.filter(
       (item)=>(item.id===props.match.params.id))[0]
     }
}

 const MapDispatchToProps =(dispatch)=>{
  return { 
    startEditExpense:(id,expense)=>dispatch(startEditExpense(id,expense)),
    startRemoveExpense: (id)=>dispatch(startRemoveExpense(id))
   }
 } 

export default connect(MapStoreToProps, MapDispatchToProps)(EditExpensePage);
//export default connect(MapStoreToProps)(EditExpensePage);