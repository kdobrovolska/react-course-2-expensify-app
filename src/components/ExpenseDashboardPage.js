import React from 'react';
import ExpenseList from './ExpenseList.js'
import ExpenseListFilters from './ExpenseListFilters.js'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList type='new'/>
    <ExpensesSummary />
  </div>
);
export default ExpenseDashboardPage;