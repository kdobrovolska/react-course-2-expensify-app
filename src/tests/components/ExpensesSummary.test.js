import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'

test('should render ExpensesSummary correctly wehis two items', ()=>{

    const wrapper=shallow(<ExpensesSummary expensesCount={2} expensesTotal={9434}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly wehis one item', ()=>{

    const wrapper=shallow(<ExpensesSummary expensesCount={1} expensesTotal={9434}/>);
    expect(wrapper).toMatchSnapshot();
});