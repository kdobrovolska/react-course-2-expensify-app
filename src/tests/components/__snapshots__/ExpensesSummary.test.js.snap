// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly wehis one item 1`] = `
<div>
  <h2>
    Expenses Count : 
    1
     
  </h2>
  <h2>
    Expenses Total : 
    $94.34
     
  </h2>
</div>
`;

exports[`should render ExpensesSummary correctly wehis two items 1`] = `
<div>
  <h2>
    Expenses Count : 
    2
     
  </h2>
  <h2>
    Expenses Total : 
    $94.34
     
  </h2>
</div>
`;
